File: core/backend/src/signals/season.ts
import { Season } from "@prisma/client";

/**
 * Classify a date's month into one of the strategy's three buckets, using
 * the supplied IANA time zone (typically Asia/Kathmandu).
 *
 * Engine v0.2:
 *   - WEAK   : March, June, November — statistically softer; deploy on MA20 dips.
 *   - STRONG : January, July, August — seasonal liquidity / Shrawan-Bhadra retail FOMO.
 *              v0.2 changed action from BUY to HOLD_FUNDS (sweep to OD loan).
 *   - NORMAL : everything else — standard weekly DCA.
 */
const WEAK_MONTHS = new Set([3, 6, 11]);
const STRONG_MONTHS = new Set([1, 7, 8]);

export function classifySeason(date: Date, timezone: string): Season {
  // Render the month in the supplied tz so e.g. an early-morning UTC date in
  // November still classifies as NOVEMBER in NPT.
  const monthStr = new Intl.DateTimeFormat("en-US", {
    timeZone: timezone,
    month: "numeric",
  }).format(date);
  const month = parseInt(monthStr, 10);
  if (WEAK_MONTHS.has(month)) return "WEAK";
  if (STRONG_MONTHS.has(month)) return "STRONG";
  return "NORMAL";
}

File: core/mobile/lib/domain/engine/season.dart
import 'types.dart';

/// Dart port of `core/backend/src/signals/season.ts`.
///
/// Classifies a date into the strategy's three seasonal buckets using the
/// supplied tz offset (default Asia/Kathmandu UTC+5:45). The engine only
/// looks at the month — picking the right tz prevents an early-morning UTC
/// date in November from being classified as October at NPT.
const _weakMonths = {3, 6, 11};
const _strongMonths = {1, 7, 8};

Season classifySeason(DateTime date, {double tzOffsetHours = 5.75}) {
  final offsetMs = (tzOffsetHours * Duration.millisecondsPerHour).round();
  final local = date.toUtc().add(Duration(milliseconds: offsetMs));
  final month = local.month;
  if (_weakMonths.contains(month)) return Season.weak;
  if (_strongMonths.contains(month)) return Season.strong;
  return Season.normal;
}
